Implement existential deposit for generic asset

aliXsed opened the pull request:
Closes #183. Keeps trace of significant assets for account ids and claim dust balances if the policy says so. Also add a migration logic for updating the newly added account store from the current storage and purging dust balances on runtime upgrade.

Commit:
Keep trace of assets' significance for account ids and do purge accounts if the policy says so

Commit:
Add doc comments

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -802,19 +815,122 @@ impl<T: Config> Module<T> {
+		// be considered a dust asset. Also any reservation or locks on the asset would mean the asset
+		// should be kept for the clearance of those operations and thus is not dust.
+		<FreeBalance<T>>::get(asset_id, who) < existential_deposit
+			&& <ReservedBalance<T>>::get(asset_id, who) == Zero::zero()
```
can reserve balance be below ED and does that matter 💭

aliXsed replied:
It can, but it could complicate the implementation. I have supposed when an account has got any reserved or locked balance, it means it still needs to exist to interact with the modules which did the reserve or lock operation. So it's important that they exist to clear up with the interactors.

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -802,19 +815,122 @@ impl<T: Config> Module<T> {
-	/// NOTE: LOW-LEVEL: This will not attempt to maintain total issuance. It is expected that
-	/// the caller will do this.
+	/// Try to mutate account of `who` in the account store based on `is_dust` for the specified asset.
+	fn try_mutate_account(asset_id: T::AssetId, who: &T::AccountId, is_dust: bool) -> Option<AccountData<T::AssetId>> {
```
nit: it's not clear what `is_dust` means in this context
would also rename function to describe what it actually does e.g. `create_or_update_account_assets`

aliXsed replied:
The naming came from similar Substrate balances functions. I will give it another thought.

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -802,19 +815,122 @@ impl<T: Config> Module<T> {
+		// should be kept for the clearance of those operations and thus is not dust.
+		<FreeBalance<T>>::get(asset_id, who) < existential_deposit
+			&& <ReservedBalance<T>>::get(asset_id, who) == Zero::zero()
+			&& Self::locks(who).is_empty()
```
would having locks on staking asset prevent dust clearing from other assets with this check?

aliXsed replied:
Dust clearing happens only when we remove an account from the account store. When an account should exist for one asset, it will be able to take care of its own dust.

aliXsed replied:
As soon as we address #185 which can be immediately after #183, then we will only check the lock for the corresponding asset.

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -185,15 +185,17 @@ pub use types::*;
+	/// Contains asset ids whose balances are greater than existential deposit.
+	significant_assets: BTreeSet<AssetId>,
+	/// True if the account should persist even with dust balances
+	persistent: bool,
```
what is the reason for adding this flag `persistent`?

aliXsed replied:
It will be useful for endowed accounts. The accounts that should exist even with dust.

jordy25519 replied:
endowed accounts aren't special, they just get some balance set via genesis config. I'm generally in favour of having less edge cases and treating all accounts the same way for simplicity sake.
if the 'persistent' account were to be cleared it doesn't change anything vs. persistent, it could be recreated by receiving a transfer, it's nonce will still exist in the system module

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -185,15 +185,17 @@ pub use types::*;
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug, Default)]
+pub struct AccountData<AssetId: Ord> {
+	/// Contains asset ids whose balances are greater than existential deposit.
+	significant_assets: BTreeSet<AssetId>,
```
would recommend changing `significant` to `active` or `existing_assets` / something related to existence

aliXsed replied:
`active` would refer to the operations happening in which an account is involved. `existing_assets` would also be confusing as other insignificant assets also exist until there's no asset with significant amounts.

aliXsed replied:
Changed it to `existing_assets` to match the new design

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -185,15 +185,17 @@ pub use types::*;
-	fn default() -> Self {
-		AccountData::None
+impl<AssetId: AtLeast32BitUnsigned + Copy> AccountData<AssetId> {
+	fn is_significant(&self) -> bool {
+		!self.significant_assets.is_empty() || self.persistent
```
assuming 'persistent' is required, makes sense to short circuit
```suggestion
	fn exists(&self) -> bool {
		self.persistent || !self.significant_assets.is_empty()
```

aliXsed replied:
Will do that.

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -213,7 +215,7 @@ pub trait Config: frame_system::Config {
 	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
 
 	/// The means of storing the balances of an account.
```
should update the doc comment to reflect its real purpose in this module

aliXsed replied:
Done

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -725,10 +740,8 @@ impl<T: Config> Module<T> {
-
 		let new_reserve_balance = b - slash;
-		Self::set_reserved_balance(asset_id, who, new_reserve_balance);
+		Self::set_balances(asset_id, who, new_free_balance, new_reserve_balance);
```
is `who` and `beneficiary` guaranteed to always be the same?
before it was:
```rust
		Self::set_free_balance(asset_id, beneficiary, new_free_balance);
		Self::set_reserved_balance(asset_id, who, new_reserve_balance);
````

aliXsed replied:
Nice catch. I will fix it right away.

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -802,19 +815,122 @@ impl<T: Config> Module<T> {
+	}
+
+	/// Remove an account's whole trace from the storage
+	fn purge(who: &T::AccountId) {
```
shouldn't purging happen for a particular asset when it's own balance goes to dust? iterating every asset in existence doesn't seem right.
when no assets are left then we could purge the AccountData storage also.

aliXsed replied:
That would be a different design. I had assumed we should purge an account and its storage only when all the assets are dust. Before that we'd only need to keep the trace of significant assets. This was based on the description in https://github.com/plugblockchain/plug-blockchain/issues/183. The design you are suggesting is less tolerant against dust, I get that. I'm open to this suggestion. Would you want me to change #183 and take this approach?

jordy25519 replied:
Ah I didn't notice this design in #183, It seems better to release storage as we're able to do so

aliXsed replied:
Done

Commit:
Transfer purged dusts to treasury

Commit:
Fix repatriate_reserved

Commit:
Purge assets individually as soon as they become insignificant

Commit:
Rename AccountData fields and functions according to the new approach in the implementation

Commit:
Rename try_mutate_account and try_mutate_account_with_dust for clarity

Commit:
Update doc comment

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -802,19 +830,114 @@ impl<T: Config> Module<T> {
+
+	/// Remove an account's whole trace from the storage and move the dusts to the treasury
+	fn purge(asset_id: T::AssetId, who: &T::AccountId) {
+		let treasury_account_id: T::AccountId = T::TreasuryModuleId::get().into_account();
```
I think it'd be nicer to pass the imbalance to a trait like `OnPurgedImbalance` or `OnDustImbalance` and let the runtime define how to handle this situation per network

aliXsed replied:
Done

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -802,19 +830,114 @@ impl<T: Config> Module<T> {
+
+	/// Call `func` wrapped in the dust check logic and according to `ExistenceRequirement` for the
+	/// `who` account. Update the account store if needed.
+	fn call_with_dust_check(asset_id: T::AssetId, who: &T::AccountId, func: impl FnOnce(), req: ExistenceRequirement) {
```
👍

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -484,6 +502,8 @@ decl_event! {
 		Minted(AssetId, AccountId, Balance),
 		/// Asset burned (asset_id, account, amount).
 		Burned(AssetId, AccountId, Balance),
+		/// Account has been deleted due to having below existential deposits
```
```suggestion
		/// Asset balance storage has been reclaimed due to falling below the existential deposit
```

Commit:
Restore the prev doc for set_free_balance and set_reserved_balance as they are now more accurate following the latest logic change

jordy25519 reviewed (commented):
quick review, still need to go through tests

Commit:
Update doc comment

Co-authored-by: Jordan <beauchjord@gmail.com>

Commit:
Move imbalance handling of dust cleaning out of generic asset.

Commit:
Only remove asset from existing_assets when you purge it as well

Commit:
Fix mock to top up treasury instead of set balance

Commit:
Just a minor improvement

Commit:
Simplify mocker

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -802,19 +827,113 @@ impl<T: Config> Module<T> {
+	) -> Option<AccountData<T::AssetId>> {
+		T::AccountStore::try_mutate_exists(who, |maybe_account| {
+			match maybe_account {
+				Some(k) => {
```
nit:
```suggestion
				Some(account) => {
```

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -484,6 +499,8 @@ decl_event! {
 		/// Asset burned (asset_id, account, amount).
 		Burned(AssetId, AccountId, Balance),
+		/// Asset balance storage has been reclaimed due to falling below the existential deposit
+		Purged(AssetId, AccountId, Balance),
```
something like this will be nicer for dapps/user
```suggestion
		DustReclaimed(AssetId, AccountId, Balance),
```

jordy25519 reviewed (commented):
generally lgtm! I don't think the persistent flag has any usefulness

Commit:
DustReclaimed

Co-authored-by: Jordan <beauchjord@gmail.com>

Commit:
Remove persistence flag

Commit:
By querying the system module itself, Test the accounts are actually added to the system module and later removed from it when purge happens

jordy25519 reviewed (commented):
looks good.
Need one more test for account transfer an amount < existential deposit should fail

Commit:
Remove the account as a separate step from mutating the account value. It's because remove can fail when there are consumers.

Also add a flag to accounts that separates the existence of an account from it's set of existing assets

jordy25519 commented on prml/generic-asset/src/lib.rs:
```diff
@@ -802,19 +820,110 @@ impl<T: Config> Module<T> {
+				Some(account) => {
+					if asset_is_dust {
+						account.existing_assets.remove(&asset_id);
+						Self::purge(asset_id, who);
+					} else {
+						account.existing_assets.insert(asset_id);
+					}
```
Looks like the system pallet can help (inspiration from here: https://github.com/paritytech/substrate/blob/3efe87daba09be18b3596e075c3733fe47bafdd5/frame/balances/src/lib.rs#L1074):
```suggestion
						Self::purge(asset_id, who);
						// GA may have the responsibility to keep an account 'alive'
						// if there are no consumers or another provider for the account exists, then we can release the storage safely
						if !account.should_exist() && (!frame_system::Pallet::<T>::is_provider_required(who) || frame_system::Pallet::<T>::providers(who) > One::one()) {
							*maybe_account = None
						}
					} else {
						account.existing_assets.insert(asset_id);
					}
```

aliXsed replied:
The problem was not just with removing the account (`*maybe_account = None`). Any change that would leave the account value to be the same as the default value would fail when a provider is required. That's why we need something like the flag `exists` that would be different than default for the case that we want to remove the last dust asset from existing_assets but leave the account for its consumer. The flag solution is efficient and better than coupling Generic Asset more and more with the system module.

jordy25519 replied:
Makes sense. Allowing account to be reclaimed by system (i.e. nonce reaped) seems like too much complexity vs. benefit, have proposed preventing it when GA is the account store in #190. Can tackle it separately if we find the need for it in future

Commit:
Add Generic Asset migration logic and test it

Commit:
Prevent an account to be reaped from the system module at this stage

jordy25519 commented on prml/generic-asset/src/tests.rs:
```diff
@@ -204,6 +209,313 @@ fn transferring_less_than_one_unit_should_fail() {
+}
+
+#[test]
+fn an_account_with_a_consumer_should_persist_in_system_even_when_ga_not_providing_it() {
```
test name should change since GA will always provide it

aliXsed replied:
> what happens when a dust amount is transferred to a non-existent account?

An account with a dust balance will be created. But we don't know if the account is going to always have a dust balance. A couple of dust balance transfer can make an account to have a significant balance. We don't need to stop that to happen. As long as the user has paid the fee for transfer we can be open to the interaction. We uniformly apply the policy to purging dusts to initiators of an interaction or a runtime upgrade.

aliXsed replied:
Done

jordy25519 reviewed (commented):
what happens when a dust amount is transferred to a non-existent account?

aliXsed commented:
> what happens when a dust amount is transferred to a non-existent account?

@holygits An account with a dust balance will be created. But we don't know if the account is going to always have a dust balance. A couple of dust balance transfer can make an account to have a significant balance. We don't need to stop that to happen. As long as the user has paid the fee for transfer we can be open to the interaction. We uniformly apply the policy of purging dusts to initiators of an interaction or a runtime upgrade.

Commit:
Update the test name

jordy25519 commented:
> @holygits An account with a dust balance will be created. But we don't know if the account is going to always have a dust balance. A couple of dust balance transfer can make an account to have a significant balance. We don't need to stop that to happen. As long as the user has paid the fee for transfer we can be open to the interaction. We uniformly apply the policy of purging dusts to initiators of an interaction or a runtime upgrade.

Can you add a test for this please

Commit:
Add a test that shows transfer_dust_balance_can_create_an_account

jordy25519 reviewed (approved):
lgtm!
